### PR TITLE
Socket communication

### PR DESCRIPTION
**Pull request by ashfurrow**
Before I get to _actually_ doing socket communication, there is some work that needs to be done first. First, and most importantly, we were inferring that an auction's slug would give us access to causality, which is not the case. After talking with @acjay, causality IDs are permanent, while slugs can – and do – change.

The solution I'm proposing is, instead of just fetching the required sale ID before loading the rest of the live auctions view controllers, let's move the fetching of the entire static data (the metaphysics query) up front. The nice things, because GraphQL is awesome, we can fetch anything we need upfront as well: sale info, sale artworks, and eventually the JWT we'll need to authenticate the websocket connection.

This simplifies a lot of other things that used to rely on the the static data coming _eventually_, but can now rely on it immediately. It removes a lot of optionality and excess properties from the sales person, too. @orta I'm keen to know what you think of these changes in particular.

Still left to do:

- [x] Remove socket.io
- [x] Add basic socket connection
- [x] Amend state reconciler class to handle new socket data format
- [x] Simulate static data loaded from JSON file (for offline UI development)
- [x] Simulate events from websockets from JSON file (for offline UI development)
- [x] Update tests

Replaces #1393.

### CODE REVIEW

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Error</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:no_entry_sign:</td>
      <td>No CHANGELOG changes made</td>
    </tr>
    
  </tbody>
</table>
  

  
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">2 Warnings</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td>PR is classed as Work in Progress</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td>Needs testing on a Phone if change is non-trivial</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

**Comment by orta** [comment]
Don't have time for an extensive review probably till next week, on a cursory glance, this looks 👍 - lots of deletions and the Sales Person being more simple is great

**Comment by ashfurrow** [comment]
No problem, this PR is pretty ambitious, probably best to wait until it's more done to get a good picture. 

Quick update: I've got stubbed metaphysics working, started on stubbing socket events (which I've started working on, just the initial lot state broadcast so far). 

screen shot 2016-04-28 at 4 48 23 pm

Still a long ways to go, but I'm feeling like we're making progress.

**Comment by ashfurrow** [comment]
Should be 🍏 now. I've marked some of the tests relating to socket communication as pending, as that code still needs to be re/written. I'll leave this up to @orta to merge or not, and I'll pick this up again on Monday regardless.

File: Artsy/Networking/ARRouter.m [hunk 1]
```diff
@@ -995,29 +995,35 @@ + (NSURLRequest *)liveSaleStaticDataRequest:(NSString *)saleID
...
+        }\
+        image {\
+          large: url(version: \"large\")\
+          thumb: url(version: \"thumb\")\
+        }\
+      }\
+    }\
+  }\
 }",
                                                  saleID];
```
Comment (orta) [inline Artsy/Networking/ARRouter.m]: These are cool for now, but we will probably need to find a better way to handle these once we have 3-4  in the router

File: Artsy/Networking/Live_Auctions/LiveAuctionSocketCommunicator.swift [hunk 2]
```diff
@@ -1,88 +1,109 @@
...
+            let url = NSURL(string: "\(host)/socket?claim_role=bidder&claim_saleId=\(saleID)&claim_userId=4C-U2DgqWh&claim_bidderId=4C-U2DgqWh")
+            return WebSocket(url: url!)
         }
     }
 }
 
 private typealias SocketSetup = LiveAuctionSocketCommunicator
 private extension SocketSetup {
 
+    class TimerCaller {
```
Comment (orta) [inline Artsy/Networking/Live_Auctions/LiveAuctionSocketCommunicator.swift]: wording feels a bit odd here,  I couldn't understand what it was  going on without reading the `setupSocket` - perhaps some simple comments above the class?

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionLotListViewController.swift [hunk 3]
```diff
@@ -51,15 +51,19 @@ class LiveAuctionLotListViewController: UICollectionViewController {
         collectionView?.registerClass(LotListCollectionViewCell.self, forCellWithReuseIdentifier: LotListCollectionViewCell.CellIdentifier)
     }
 
+    override func viewDidAppear(animated: Bool) {
+        super.viewDidAppear(animated)
+    }
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionLotListViewController.swift]: superfluous

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift [hunk 4]
```diff
@@ -5,61 +5,51 @@ import Interstellar
...
+    let auctionViewModel: LiveAuctionViewModelType
+    var pageControllerDelegate: LiveAuctionPageControllerDelegate?
 
-    private(set) var lots = [LiveAuctionLotViewModelType]()
     private let stateManager: LiveAuctionStateManager
 
-    // Lot currentloy being looked at by the user.
-    var currentFocusedLot = Observable<Int>()
+    // Lot currently being looked at by the user.
+    var currentFocusedLotID = Observable<Int>()
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift]: 👍

File: Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift [hunk 5]
```diff
@@ -17,6 +17,7 @@ protocol LiveAuctionLotViewModelType: class {
     var estimateString: String { get }
     var lotPremium: String { get }
     var lotName: String { get }
+    var lotId: String { get }
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift]: `lotID`? ( it's used in other places with caps ( `model.liveAuctionLotID` ) )

**Comment by orta** [comment]
Feels super nice 👍- great work here

File: Artsy/Networking/ARRouter.m (hunk 1 above)
Comment (ashfurrow) [inline Artsy/Networking/ARRouter.m]: Agreed – this is the only one involved with Live Auctions. I think @alloy has plans for broader metaphysics architecture.

File: Artsy/Networking/Live_Auctions/LiveAuctionSocketCommunicator.swift (hunk 2 above)
Comment (ashfurrow) [inline Artsy/Networking/Live_Auctions/LiveAuctionSocketCommunicator.swift]: Good idea.

**Comment by ashfurrow** [comment]
Thanks! Today I'll clean it up and add some of the pending tests back in (and fix their related code).

**Comment by ashfurrow** [comment]
Okay, got this wrapped up. Next steps will be updating individual bits based on different events, but that can be another PR. We need to look for any places that are using hard-coded values that we had for testing the UI, but need to be replaced with actual values now – hopefully there's nothing in the UI we're not already parsing. I could you some help on this one if you have time, Orta. 

@orta your tests were very helpful making sure I had corner-cases of JSON parsing covered :100: I like how this is shaping up.

File: Artsy/Models/API_Models/Live/LiveEvent.h [hunk 6]
```diff
@@ -22,8 +26,8 @@ typedef NS_ENUM(NSInteger, LiveEventType) {
 // We do this to expose these to its children
 
 @property (nonatomic, assign, readonly) NSInteger amountCents;
-@property (nonatomic, copy, readonly) NSString *source;
-@property (nonatomic, assign, readonly) BOOL isConfirmed;
+@property (nonatomic, strong, readonly) LiveBidder *_Nullable bidder;
+@property (nonatomic, strong, readonly) NSString *sourceOrDefaultString;
```
Comment (orta) [inline Artsy/Models/API_Models/Live/LiveEvent.h]: minor niggle - strings should be `copy` as they have mutable subclasses, you know you're dealing with unmutable data this way

File: Artsy/Networking/Live_Auctions/StarscreamExtensions.swift [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    }
+
+    func disconnect() {
+        disconnect(forceTimeout: nil)
+    }
+
+    func writePing() {
+        writePing("2".dataUsingEncoding(NSUTF8StringEncoding)!)
+    }
+}
```
Comment (orta) [inline Artsy/Networking/Live_Auctions/StarscreamExtensions.swift]: 👍

File: Artsy/View_Controllers/Live_Auctions/LiveActionStateReconciler.swift [hunk 8]
```diff
@@ -24,160 +24,85 @@ State update includes:
...
-                }
-                .flatMap { eventJSON in
-                    return LiveEvent(JSON: eventJSON)
-                }
-
-            viewModel.addEvents(newEvents)
-        }
+        // OK, this looks weird. Let's unpack.
+        // derivedState["askingPriceCents"] is an AnyObject?, and casting it conditionally to a UInt64 always fails.
+        // Instead, we'll use the UInt64(_ text: String) initialzer, which means we need to unwrap the AnyObject? and
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveActionStateReconciler.swift]: typo: `initialzer`

**Comment by orta** [comment]
OK, let's do this 👍

File: Artsy/Models/API_Models/Live/LiveEvent.h (hunk 6 above)
Comment (ashfurrow) [inline Artsy/Models/API_Models/Live/LiveEvent.h]: Right, makes sense. Forgot, as this is a computed property.
